cpp/test/distance/masked_nn.cu

```diff
@@ -0,0 +1,500 @@
+
+    bool init_out = true;
+    ParamT masked_l2_params{RedOpT{}, PairRedOpT{}, Sqrt, init_out};
+
+    int m          = params.m;
+    int n          = params.n;
+    int k          = params.k;
+    int num_groups = params.num_groups;
+
+    auto x_view          = raft::make_device_matrix_view(x.data(), m, k);
+    auto y_view          = raft::make_device_matrix_view(y.data(), n, k);
+    auto x_norm          = raft::make_device_vector_view(xn.data(), m);
+    auto y_norm          = raft::make_device_vector_view(yn.data(), n);
+    auto adj_view        = raft::make_device_matrix_view(adj.data(), m, num_groups);
+    auto group_idxs_view = raft::make_device_vector_view(group_idxs.data(), num_groups);
+    auto out_view        = raft::make_device_vector_view(out, m);
+
+    maskedL2NN<DataT, OutT, IdxT>(handle,
+                                  masked_l2_params,
+                                  x_view,
+                                  y_view,
+                                  x_norm,
+                                  y_norm,
+                                  adj_view,
+                                  group_idxs_view,
+                                  out_view);
+
+    handle.sync_stream(stream);
+  }
+};
+
+template <typename T>
+struct CompareApproxAbsKVP {
+  typedef typename raft::KeyValuePair<int, T> KVP;
+  CompareApproxAbsKVP(T eps_) : eps(eps_) {}
+  bool operator()(const KVP& a, const KVP& b) const
+  {
+    T diff  = raft::abs(raft::abs(a.value) - raft::abs(b.value));
+    T m     = std::max(raft::abs(a.value), raft::abs(b.value));
+    T ratio = m >= eps ? diff / m : diff;
+    return (ratio <= eps);
+  }
+
+ private:
+  T eps;
+};
+
+template <typename T>
+struct CompareExactKVP {
+  typedef typename raft::KeyValuePair<int, T> KVP;
+  bool operator()(const KVP& a, const KVP& b) const
+  {
+    if (a.value != b.value) return false;
+    return true;
+  }
+};
+
+template <typename K, typename V, typename L>
+::testing::AssertionResult devArrMatch(const raft::KeyValuePair<K, V>* expected,
+                                       const raft::KeyValuePair<K, V>* actual,
+                                       size_t size,
+                                       L eq_compare,
+                                       cudaStream_t stream = 0)
+{
+  typedef typename raft::KeyValuePair<K, V> KVP;
+  std::shared_ptr<KVP> exp_h(new KVP[size]);
+  std::shared_ptr<KVP> act_h(new KVP[size]);
+  raft::update_host<KVP>(exp_h.get(), expected, size, stream);
+  raft::update_host<KVP>(act_h.get(), actual, size, stream);
+  RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  for (size_t i(0); i < size; ++i) {
+    auto exp = exp_h.get()[i];
+    auto act = act_h.get()[i];
+    if (!eq_compare(exp, act)) {
+      return ::testing::AssertionFailure()
+             << "actual=" << act.key << "," << act.value << " != expected=" << exp.key << ","
+             << exp.value << " @" << i;
+    }
+  }
+  return ::testing::AssertionSuccess();
+}
```

Comment:
These helper functions seem to be the same as the ones used by the fused_l2_nn test. Please remove code duplication by factoring them to a common test utils file. (Possibly a subtask for #1171).

Reply:
I have added this as a subtask of #1171 and have removed `CompareExactKVP` as it was not used in maskedL2NN tests.